### PR TITLE
Inline mirror://sourceforge

### PR DESCRIPTION
**Pull request by mgorny**
Sourceforge is using download bouncer these days, so tpm entry lists only 1 server. Inline it.

### CODE REVIEW

**Comment by Whissi** [comment]
What's the problem you are going to solve? Sure, sourceforge has currently only one server entry in mirror file. But do we know if this URL is stable? If not, do we want to mass-change ebuilds again? My point is, why don't we just keep current status which will allow us to update all ebuilds using sourceforge using a single file. Does current status hurt anyone?

**Comment by mgorny** [comment]
> What's the problem you are going to solve? Sure, sourceforge has currently only one server entry in mirror file. But do we know if this URL is stable?

Yes, it is backwards-compatible. But it isn't correct. Removing the obfuscation is the first step towards fixing that.

> If not, do we want to mass-change ebuilds again?

Yes, technically the majority of ebuilds need to be changed 'again'. And yes, if sourceforge ever changes URLs again, we'd have to change ebuilds. Just like with any other hosting.

> My point is, why don't we just keep current status which will allow us to update all ebuilds using sourceforge using a single file. Does current status hurt anyone?

Do you have any guarantee that the URIs will change in a compatible way permitting reuse of mirror? So far faking stuff via mirror:// only resulted in spread of invalid URIs.

**Comment by floppym** [comment]
Or we could just stick with the backward-compatible URLs that have been working for 10+ years? I agree that this change seems quote pointless.

**Comment by mattst88** [comment]
But... what's the rationale? What does removing mirror://sourceforge accomplish? What goal does it get you closer to?

**Comment by chewi** [comment]
I haven't looked in detail but I trust you got these right. At least for the games, it's not the end of the world if a couple are wrong as we can easily fix it later.

I don't really care too much either way whether this goes through although I do agree that it is very easy to get SourceForge URLs wrong.

**Comment by mgorny** [comment]
> But... what's the rationale? What does removing mirror://sourceforge accomplish? What goal does it get you closer to?

Making things more straightforward. This indirection is useless and causes extra work for everyone.

If someone writes a new ebuild, and notices sourceforge in thirdpartymirrors, instead of copying and updating the correct sourceforge download URL, he wastes his time figuring out how to apply `mirror://sourceforge`. Except that has no benefit, and most likely causes him to get the URL wrong which causes permanent redirect for our users.

This is self-propelling problem. As long as the useless mirror entry sits there, people will notice it and switch their ebuilds to it, causing even more people to use that in the future. All for no benefit, all causing extra work in the future and more incorrect URLs in ebuilds. Of course, we can hopefully assume upstream will keep them redirecting forever.

Compared to that, the cost of one-time machine update of all SRC_URIs is low. And since the entry disappears, people stop using it in new ebuilds.

**Comment by atomsymbol** [comment]
> > But... what's the rationale? What does removing mirror://sourceforge accomplish? What goal does it get you closer to?
> 
> Making things more straightforward. This indirection is useless and causes extra work for everyone.
> 
> If someone writes a new ebuild, and notices sourceforge in thirdpartymirrors, instead of copying and updating the correct sourceforge download URL, he wastes his time figuring out how to apply `mirror://sourceforge`. Except that has no benefit, and most likely causes him to get the URL wrong which causes permanent redirect for our users.
> 
> This is self-propelling problem. As long as the useless mirror entry sits there, people will notice it and switch their ebuilds to it, causing even more people to use that in the future. All for no benefit, all causing extra work in the future and more incorrect URLs in ebuilds. Of course, we can hopefully assume upstream will keep them redirecting forever.
> 
> Compared to that, the cost of one-time machine update of all SRC_URIs is low. And since the entry disappears, people stop using it in new ebuilds.

There is also the possibility [theoretical at least] of automatically converting **all** links to a desired format when creating individual ebuilds, via a mandatory ebuild tool that the ebuild maintainer would have to run. In case the tool has a rule in place to recognize the link initially provided by the maintainer, the initial form of the link can be the one that is the most convenient and fastest to the maintainer. It might be an addition to `repoman` inspired by e.g. Go's [gofmt](https://golang.org/cmd/gofmt/) and [gofix](https://golang.org/cmd/fix/).

**Comment by orlitzky** [comment]
Most of the entries in thirdpartymirrors are there for cargo-cult reasons. For many years there was no documentation about the thirdpartymirrors file. This led to cargo-culting: people saw the word "mirrors," and thought, "ooh there are some mirrors for my distfile that I could put in there!" Which they then did, and immediately forgot about -- leading to a pile of dead URLs in the thirdpartymirrors file.

But that was never the intention. Anything that's not fetch/mirror-restricted is automatically mirrored onto Gentoo infrastructure without any user/developer interaction (or maintenance) at all. The thirdpartymirrors feature was intended to mirror the other stuff, things that can't be mirrored onto Gentoo infrastructure. It doesn't make sense to add **everything** to the thirdpartymirrors file because it's so awful to maintain, and there's basically zero benefit.

So, yes, please kill sourceforge from that file because it serves no purpose to begin with. Everything on SF is mirror/fetchable.

**Comment by Whissi** [comment]
> So, yes, please kill sourceforge from that file because it serves no purpose to begin with. Everything on SF is mirror/fetchable.

I think you are mixing things. It's not about allowed to mirror/fetch. In the past, SF had no download 'bouncer', there were no single "wgetable" URL to point to. So due to their File Release System (FRS), you had to call exact mirror URL (`http://<mirror>.dl.sourceforge.net`). That's why we had all SF mirrors listed in https://sources.gentoo.org/cgi-bin/viewvc.cgi/gentoo-x86/profiles/thirdpartymirrors?view=markup

The new owner has changed this and introduced a 'wgetable' bouncer ([probably because of HTTPS](https://sourceforge.net/blog/small-but-significant-changes-to-project-pages/)).

**Comment by orlitzky** [comment]
> > So, yes, please kill sourceforge from that file because it serves no purpose to begin with. Everything on SF is mirror/fetchable.
> 
> I think you are mixing things. It's not about allowed to mirror/fetch. In the past, SF had no download 'bouncer', there were no single "wgetable" URL to point to.

It wasn't easy to find, but if you right-clicked on their "your download will start in five seconds..." link, you could get a real URL to the file. Sure you're just picking one non-canonical mirror "at random," but that's what we do with e.g. PHP too. Once that file gets mirrored by Gentoo, it doesn't matter any more. The benefit to adding things in thirdpartymirrors is next to zero, but the cost (in maintenance, see e.g. how much time we just spent responding to each other in this PR =) is high.


> The new owner has changed this and introduced a 'wgetable' bouncer.

Thankfully... Despite SF's unusual past, the existing mirror entry is kinda pointless. And for almost everybody else in thirdpartymirrors who never made it hard to get at the underlying URL in the first place, being listed in thirdpartymirrors was always pointless.

The Gentoo mirrors are just an all-around better solution to this problem, when they can be used (when no fetch/mirror restrictions exist). A huge number of thirdpartymirrors entries are completely broken and nobody has complained or has had to fix them, because they didn't need to be there in the first place =)

**Comment by Whissi** [comment]
I agree that once file landed on Gentoo mirrors, SRC_URI isn't really relevant anymore. However, tell that our users, see https://bugs.gentoo.org/661178 ;-)

Anyway, the point was: I (and other) still don't understand the need for mass editing so many packages. Current thirdpartmirrors entry for SF is valid and will give us a small chance to update most ebuilds with one edit if only host should change in future. Sure, if they are going to break url structure after 15+ years, this won't work. But why do two mass edits? Just wait for that day and keep status quo which is working and doesn't hurt anyone.

**Comment by rafaelmartins** [comment]
I already muted this thread twice, and keep getting re-added to it by this bot every time it runs the tests again, please fix this bot

**Comment by bella-forever** [comment]
EtherApe-INFO: 22:03:36.143: sctp protocol not supported
EtherApe-INFO: 22:03:36.144: ddp protocol not supported
EtherApe-INFO: 22:03:36.144: ddp protocol not supported
EtherApe-INFO: 22:03:36.144: ddp protocol not supported
EtherApe-INFO: 22:03:36.144: ddp protocol not supported
EtherApe-INFO: 22:03:36.167: get_interface result: ''
EtherApe-INFO: 22:03:36.170: Available interfaces for capture: nfqueue nflog eth0 wlan1 any lo wlan0
EtherApe-INFO: 22:03:36.239: Link type is Ethernet
EtherApe-INFO: 22:03:36.244: Diagram started
EtherApe-INFO: 22:03:37.739: New node: IP: 192.168.1.10. Number of nodes 1
EtherApe-INFO: 22:03:37.740: New node: IP: 224.0.0.2. Number of nodes 2
EtherApe-INFO: 22:03:37.741: New node: IP: 192.168.1.11. Number of nodes 3
EtherApe-INFO: 22:03:37.741: New node: IP: 34.205.201.140. Number of nodes 4
unexpected EOF in read_all()
critical: read_all() failed on control socket
Segmentation fault



helppppppppppp plllllwwwwwwww

**Comment by dermoench42** [comment]
<app-editors/scite/scite-4.1.3 removed in early january, no further adjustments required, there is no 'mirror' hint in current available ebuilds.

**Comment by scardracs** [comment]
I think drop the mirror:// for sourceforge could be a good idea as bouncer select the best server, depending on position, instead of forcing 1 mirror for everyone. When I'm pushing PRs I prefer to use the https anyway, also because pkgdev starts to show ForceRedirect as warning

**Comment by antecrescent** [comment]
What's stopping us from either closing or merging this PR?

Currently, the ebuilds affected by this PR do not follow the [devmanual's recommendation](https://devmanual.gentoo.org/ebuild-writing/variables/#third-party-mirrors) on when to use `thirdpartymirrors`.
> There are two valid cases for using thirdpartymirrors:
> 1. providing multiple download locations for mirror- or fetch-restricted packages,

According to https://github.com/gentoo/gentoo/pull/13105#issuecomment-536999242, everything on SF is mirror-/fetchable. Even if this was not the case, the current entry for SF is just the upstream bouncer `https://downloads.sourceforge.net`. Effectively, we don't "provide multiple download locations" but add an unnecessary level of indirection to the existing ebuilds.

> 2. dealing with upstreams that distribute their distfiles via a network of mirrors without a primary download location or a bouncer service.

SF has provided the aforementioned bouncer service for at least 5 years.

I do concede that the devmanual isn't as authoritative as the PMS, but it's still a widely used resource among ebuild authors.
In short, I believe that achieving consistency with the devmanual is reason enough to merge this PR. 
If we decide not to merge this PR, we should probably rework `thirdpartymirrors` and amend the devmanual's section on it.

One last remark regarding https://github.com/gentoo/gentoo/pull/13105#issuecomment-537154284:
> Current thirdpartmirrors entry for SF is valid and will give us a small chance to update most ebuilds with one edit if only host should change in future. 

While I do see the merit of the ability to mass-update sources with a simple edit, that discussion should - in my opinion -  be held over at gentoo-dev, independently of the PR at hand.

**Comment by orlitzky** [comment]
I didn't know (or forgot that) there was a PR open for this. I "recently" cleaned up the cran, gmt, ubuntu, and samba mirrors. @DerDakon did the same for qmail.

It's pointless indirection abusing a feature intended for something else, and the list of mirrors is perpetually bitrotten.

**Comment by floppym** [comment]
I rescind my earlier objection. If someone wants to rebase this, go for it.

**Comment by antecrescent** [comment]
@mattst88 @Whissi Do you still object? I think, it would be great, if we finally merged (or closed) this PR after five years.
Several other mirrors have apparently been cleaned up since then (https://github.com/gentoo/gentoo/pull/13105#issuecomment-2041141819).

**Comment by mattst88** [comment]
No objection from me. (Whissi is retired)

**Comment by antecrescent** [comment]
@mgorny All participants on this thread are in favor of merging this PR now :)

**Comment by mgorny** [comment]
> @mgorny All remaining participants on this thread are in favor of merging this PR now :)

Do you feel like updating it / making a new PR? It obviously doesn't apply 5 years later. And I definitely don't have the whatever invocation I've used to make it anymore.

**Comment by antecrescent** [comment]
Yes sure!

**Comment by antecrescent** [comment]
This can be closed now (d8c27f274538c324fa458dcb17eba48636d34acc) ^^ Thanks everyone!
